app/src/main/java/com/example/antitheft/MainActivity.java

```diff
@@ -0,0 +1,174 @@
+        sendButton = findViewById(R.id.sendButton);
+        activity = this;
+
+        sendButton.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                boolean permissionsGranted = checkSelfPermission(Manifest.permission.SEND_SMS) == PERMISSION_GRANTED;
+                if (permissionsGranted) {
+                    String telephoneNumber = ((EditText) findViewById(R.id.telephoneNumber)).getText().toString();
+                    // Checking the peer.
+                    SMSPeer peer = new SMSPeer(telephoneNumber);
+                    if (peer == null)
+                        Toast.makeText(context, R.string.toast_non_valid_peer, Toast.LENGTH_SHORT).show();
+                        // If peer is not null, it sends the command.
+                    else sendCommand(new SMSMessage(peer, FULL_DEFAULT_LOCATE_COMMAND));
+                } else {
+                    // If SMS permissions are not granted, it will show a toast, because
+                    // the app can't send commands without that permission.
+                    Toast.makeText(context, R.string.toast_app_needs_SMS_permissions, Toast.LENGTH_SHORT).show();
+                    requestPermissions(new String[]{Manifest.permission.SEND_SMS}, REQUEST_CODE);
+                }
+            }
+        });
+        // Request permissions.
+        requestPermissions(PERMISSIONS, REQUEST_CODE);
+
+        // Immediately prepare the last known location.
+        client = LocationServices.getFusedLocationProviderClient(this);
+        // Callback to retrieve the last known location.
+        client.getLastLocation().addOnSuccessListener(this, new OnSuccessListener<Location>() {
+            @Override
+            public void onSuccess(Location location) {
+                locationParser.setLocation(location);
+            }
+        });
+
+    }
+
+    /**
+     * This method sends the command, calling the method in {@link GPSCommandHandler} class, if
+     * command is not null.
+     *
+     * @param smsMessage The message to send.
+     * @see com.eis.smslibrary.SMSMessage
+     * @see GPSCommandHandler
+     * @see GPSCommandHandler#sendMessage(SMSMessage)
+     */
+    private void sendCommand(@Nullable SMSMessage smsMessage) {
+        if (smsMessage != null)
+            new GPSCommandHandler().sendMessage(smsMessage);
+        else
+            Toast.makeText(this, R.string.sendCommand_fail, Toast.LENGTH_SHORT).show();
+    }
+
+    /**
+     * This method retrieves the last known {@link Location}, parsed as String.
+     *
+     * @return The last known location.
+     * @see Location
+     * @see LocationParser
+     * @see LocationParser#toString()
+     */
+    public static String getCurrentLocation() {
+        String locationTag = "LOCATION";
+        // Callback to retrieve the last known location.
+        client.getLastLocation().addOnSuccessListener(activity, new OnSuccessListener<Location>() {
```

Comment:
You could remove the `private Activity activity` variable and use `this` directly above all because if you need a `MainActivity`'s method you don't have it in a normal `Activity` and then you can always use `this` to invoke the class itself.
Did you need that variable for anything else?

```suggestion
        client.getLastLocation().addOnSuccessListener(this, new OnSuccessListener<Location>() {
```

Reply:
Well, that's the main issue...
I can't use `this` in a static method.
And, unfortunately, not even a Context is enough to add a listener.

But thank you for spotting that. I'm gonna add more @ see references in next commit.

Reply:
I think I improved this now...
I also realized that Activity instance isn't mandatory, and it still works!